src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -319,7 +323,27 @@ public void checkValidState() {
 
     // Helper methods
 
-    public abstract long maxAlignMask();
+    @ForceInline
```

Comment:
When adding `maxAlign` I have played with something similar as what you have done. My recollection is that doing this doesn't 100% solve the issue, as it's possible to have polluted profile for `maxAlignMask` (e.g. the JVM will bias the method implementation towards the 1-2 layouts that seem to be more common, and put everything else in an uncommon branch). Maybe what you see is that, since the virtual call is gone, there is still a net gain.

Does the problem only manifest with bulk copy? Or also with plain memory access? If the former, perhaps some other inlining issue with bulk copy (which is a big method) could be at play.

Reply:
I just checked it with vector operations only, there's check in `ScopedMemoryAccess` `loadFromMemorySegment` which checks max align in case of polluted access this will create polymorphic call.

- However when I look right now I think there's second option. Instead of changing `maxAlign`, the check in `loadFromMemorySegment`, could be replaced by checking if segment instance is native or `Bytes`.?

Reply:
is this the affected code?
```
// @@@ Smarter alignment checking if accessing heap segment backing non-byte[] array
        if (msp.maxAlignMask() > 1) {
            throw new IllegalArgumentException();
        }
 ```

Reply:
Exactly!

Reply:
I guess which solution is taken depends on how deep of a problem this is. I suggest writing a benchmark using e.g. bulk copy between segments, and see if the same issue occurs. We do have tests for polluted memory segment access and we have not seen issues with maxAlignMask showing up there, so I'm curious as to why the vector use-site seems to be more problematic. I wonder if part of the issue is the lack of argument type profiling - e.g. profile info based on the type of arguments in a static call. I think Method/VarHandle have that enabled by default (for obvious reasons) and there are hacks in the JVM to allow that for Unsafe as well. But that support is not enabled for this particular vector call, and I wonder if that could lead to the problem you describe.

https://github.com/openjdk/jdk/blob/master/src/hotspot/share/oops/methodData.cpp#L1583

Reply:
I created test with core method like this
```java
    @CompilerControl(DONT_INLINE)
    private static long sum(MemorySegment in) {
        long sz = in.byteSize() / Integer.BYTES;
        long result = 0L;
        for (long i = 0; i < sz; i += Integer.BYTES) {
            result += (int) JAVA_INT.get(in, i);
        }

        return result;
    }
```
But results after / before patch does not differs so much. 

Interesting I see some additional optimization turns on after some time (it should be fully JITed during warm-up)
```
Iteration   3: 228720.552 ns/op
Iteration   4: 220640.677 ns/op
Iteration   5: 12248.776 ns/op
Iteration   6: 9470.916 ns/op
Iteration   7: 9422.982 ns/op
```

Later I can try a bit with `TypeProfileLevel` looking at code setting it to `2` should enable mentioned profiling always.

Reply:
@mcimadamore,

I've updated JVM method to allow profiling arguments as you hinted, for vector operations
```
bool MethodData::profile_unsafe(const methodHandle& m, int bci) {
  Bytecode_invoke inv(m , bci);
  if (inv.is_invokevirtual()) {
    Symbol* klass = inv.klass();
    if (klass == vmSymbols::jdk_internal_misc_Unsafe() ||
        klass == vmSymbols::sun_misc_Unsafe() ||
        klass == vmSymbols::jdk_internal_misc_ScopedMemoryAccess()) {
      Symbol* name = inv.name();
      if (name->starts_with("get") || name->starts_with("put") || name->starts_with("load") || name->starts_with("store")) {
        return true;
      }
    }
  }
  return false;
}
```
This improved performance with benchmarks, in same way as a patch:

Before
```
MixedAccessBenchmarks.directCopy          1048576  avgt   10    22763.442 ±   2145.833  ns/op
MixedAccessBenchmarks.pollutedAccessCopy  1048576  avgt   10  3536082.952 ± 309959.183  ns/op
```
After
```
Benchmark                                  (size)  Mode  Cnt       Score      Error  Units
MixedAccessBenchmarks.directCopy          1048576  avgt   10   23006.005 ± 2497.308  ns/op
MixedAccessBenchmarks.pollutedAccessCopy  1048576  avgt   10  254049.440 ± 5218.658  ns/op
```

Reply:
Thanks for the experiment - that seems a better way to handle this situation, but I'll leave further comments to @PaulSandoz or @iwanowww.

Reply:
Expanding the profiling to for method names starting with "loadFrom" and "storeInto" seems reasonable, given the most other access methods are also profiled (although not all the read-modify-write methods are).

I think we should separate out the comparison of method names thereby we don't inadvertently profile other methods just in case new methods get added. Plus we can add some comments to the class doc of `Unsafe` and `ScopedMemoryAccess` to indicate such methods are profiled. Ideally I would prefer an HotSpot-specific annotation indicated that all arguments of an annotated method should be profiled.
 
The profiling of all arguments is a blunter hammer that necessary in this case but so be it.

Reply:
Adding annotation for this could be useful for profiling other methods too. I'll try to find some time and prepare some draft.

One thing I see that adding such annotation could consume one bit here (still there will be 5 spare): 
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/oops/method.hpp#L98

Reply:
Ok. I think it may be better to propose the more focused fix to JDK 20, and then use that PR as the opportunity to ask the question to the HotSpot devs whether they think an annotation would be useful before diving in.

Reply:
So you are suggesting to create PR with annotations and ask on hotspot-dev what they think about it?

Reply:
No, i am suggesting to create a PR with changes to https://github.com/openjdk/jdk/blob/master/src/hotspot/share/oops/methodData.cpp#L1583 as you proposed. Then before you dive into any annotation driven approach ask on the hotspot email list, or ask in that PR.